### PR TITLE
feat!: let DependencyProvider prioritize dependencies

### LINKED ISSUE #18: Don't require owned dependencies from provider
Is there a way to avoid the clones and allocation in https://github.com/mpizenberg/elm-test-rs/blob/31596f874a6e9340350de3e7c783a5cde9ca396b/src/solver.rs#L439? In Cargo we use `Rc<Vec<Summary>>` https://github.com/rust-lang/cargo/blob/07162dbaa84aaad97d6f3be5069265e784179bcd/src/cargo/core/resolver/dep_cache.rs#L35, so that a cache hit is  and Rc clone instead of a Vec clone. I wonder if there is a trate bound that works for the algorithm and allows returning `Vec<>`, `Rc<Vec<_>>`, `im::Vector<_>`, or even a leaked `&'static [_]`?

Originally brought up in https://github.com/mpizenberg/pubgrub-rs/pull/15#issuecomment-704308551

### PR DESCRIPTION
**Pull request by Eh2406**
This follows up on #40. 

The first commit explores `pick_package` approach as described in https://github.com/pubgrub-rs/pubgrub/issues/40#issuecomment-714865685.
The second commit explores the `make_decision` approach.
The third commit uses a BTreeMap to remove Hash req on `OfflineDependencyProvider`

Beadback is definitely appreciated.

Edit: somewhat accidentally closes #18

### CODE REVIEW

File: src/solver.rs
Comment (mpizenberg) [inline src/solver.rs]: That's a really cool trick!

File: src/solver.rs
Comment (mpizenberg) [inline src/solver.rs]: Is it `potential_packages` instead of `get_dependencies`?

File: src/solver.rs
Comment (mpizenberg) [inline src/solver.rs:133]: shouldn't be the reverse? the version is supposed to be contained in the intersection of terms.

**Review by mpizenberg — COMMENTED** [review]
I personally like the complete control given back to the dependency provider with `make_decision`. The usage of the BTree is also a good idea to remove the `hash` constraint. The trick with `Borrow` to enforce a correct package is also really nice.

There are few oddities though in package docs. Names that do not make a lot of sense anymore like the "reverse" terminology for the dependency provider. And a few other things that I found weird at first read like the error when the version is in the term (should be the opposite?).

**Comment by Eh2406** [comment]
I force pushed to fix the commit names. I also addressed @mpizenberg's comments. I have 2 remaining concerns:

1. The module documentation still refers to `list_available_versions`. I could use some help describing the new method.
2. I added a new trait `Constraints` to hide `Term` is this a good idea? Alternatively we can make `Term` public with only the one method available, or as we are filtering to positive terms we can convert to a `Range` which is already public.

**Comment by mpizenberg** [comment]
> Alternatively we can make Term public with only the one method available, or as we are filtering to positive terms we can convert to a Range which is already public

Good point! Since potential packages are only picked if there is at least one positive derivation, the intersection is necessarily a positive term. Using a `Range` in the public argument type would make more sense from the point of view of the dependency provider!

**Comment by Eh2406** [comment]
switched to `Range`. I (at this time of night) could not get the lifetimes to work for `&Range`  so used `Borrow<Range<V>>`

File: src/solver.rs
Comment (Eh2406) [inline src/solver.rs]: The borrow checker did not like using a match on iter. This `if ... is_none ... drop` is a workaround. I believe this is a case where [`Polonius`](https://github.com/rust-lang/polonius) could help, when and if it lands in rustc.

File: src/solver.rs
Comment (mpizenberg) [inline src/solver.rs]: What about `potential_packages = state.partial_solution.potential_packages();` for the name instead of `iter`?

**Review by mpizenberg — APPROVED** [review]
Looks good to me, with maybe some changes to some doc comments. Waiting on others reviews

File: src/solver.rs
Comment (Eh2406) [reply src/solver.rs]: good point will do.

**Comment by Eh2406** [comment]
> with maybe some changes to some doc comments

What did you have in mind? Also:

> The module documentation still refers to list_available_versions. I could use some help describing the new method.

**Comment by aleksator** [comment]
Hopefully I'll be able to look at PRs tomorrow or the day after.

File: src/solver.rs
Comment (aleksator) [inline src/solver.rs:257]: I propose to provide a default implementation for this method with the heuristic we were using previously.

This way we still give control to the users who want it, but provide the option to leave the way we pick packages as the library suggests.

File: src/solver.rs
Comment (aleksator) [inline src/solver.rs]: Should we return a newly introduced error here?

File: tests/proptest.rs
Comment (aleksator) [inline tests/proptest.rs]: Isn't this `DepedencyProvider` rather new? :smile: 

Maybe `OldestVersionsDependencyProvider` would be better. Alternatively we leave it as `ReverseDependencyProvider` if we go with the suggestion to return newest versions first as the default trait implementation.

File: src/solver.rs
Comment (Eh2406) [reply src/solver.rs:257]: Do you want to keep the `list_available_versions`? Because without that I don't know how to build a default implementation. It was removed as we otherwise don't need it.

**Comment by aleksator** [comment]
> This does simplify the code, but the performance characteristics are different. In the first case, we only need to browse terms for any positive before computing intersection. In the refactor we always compute the intersection. If benchmarking showcases similar performances, it's worth it. Otherwise let's leave it for another round of performance adjustments.

On the other hand, we don't iterate over not intersected derivations every time.

I run the benchmarks a few times before submitting the change.
Before the change it has completed in 1.65ms on my machine, after the change in 1.67ms. I'm not sure if this is spurious or not, since when I benchmarked something else before I was getting much larger variations without any code changes.

File: src/solver.rs
Comment (Eh2406) [reply src/solver.rs]: or just panic, it is not clear to me how a user of our code could use a `match` in this case. Like the only fix is to fix there implementation.

File: src/solver.rs
Comment (mpizenberg) [reply src/solver.rs:257]: I don't think we should keep `list_available_versions` in the interface. Maybe we could have a free function (or collection of free functions) that could help. Something like `pick_first_strategy`, `pick_lowest_number_of_versions_strategy` etc. The names are terrible, but that's just for the sake of the example

File: src/solver.rs
Comment (mpizenberg) [reply src/solver.rs]: I'm not sure what panic brings. With an error, the caller can do whatever they want when a `Failure` happens (they could panic). And new error wouldn't bring much either since they have to fix their bug anyway.

**Comment by Eh2406** [comment]
When I ran the benchmarks from #34, criterion was fairly sure there is a small regression comparing ecaeb390311297b478e73945290c304ed086d074 with acc391ec75876e5dfaa19216510612f792a8d881. It is made smaller if we eagerly do the intersections and go ahead and do away with `derivations_not_intersected_yet` entirely. But criterion thinks it is still a 3-5% regression.
It is a code complexity vs perf trade off, I'd be ok either way.

**Comment by aleksator** [comment]
> When I ran the benchmarks from #34, criterion was fairly sure there is a small regression comparing [ecaeb39](https://github.com/pubgrub-rs/pubgrub/commit/ecaeb390311297b478e73945290c304ed086d074) with [acc391e](https://github.com/pubgrub-rs/pubgrub/commit/acc391ec75876e5dfaa19216510612f792a8d881). It is made smaller if we eagerly do the intersections and go ahead and do away with `derivations_not_intersected_yet` entirely. But criterion thinks it is still a 3-5% regression.
> It is a code complexity vs perf trade off, I'd be ok either way.

I've restored "if conditions" and still simplified the code a little with an early return.

File: src/solver.rs
Comment (Eh2406) [reply src/solver.rs:257]: I was sceptical, but gave it a try. And it worked out really well! Take a look and let me know what you think.

File: src/lib.rs
Comment (mpizenberg) [inline src/lib.rs]: I'd write something like this:
```txt
The first method [make_decision](...) chooses a package and available version
compatible with the provided options.
A helper function [make_fewest_...](...) is provided for convenience
in cases when lists of available versions for packages are easily obtained.
The strategy of that helper function consists in choosing the package
with the fewest number of compatible versions to speed up resolution.
But in general you are free to employ whatever strategy suits you best
to pick a package and a version.
```

File: tests/proptest.rs
Comment (mpizenberg) [inline tests/proptest.rs]: Like the factorization!

**Review by mpizenberg — APPROVED** [review]

File: src/lib.rs
Comment (Eh2406) [reply src/lib.rs]: Great wordsmithing! incorporated.

**Comment by mpizenberg** [comment]
The last commit makes the code cleaner by using `min_by` but I think it doubles the amount of calls to `list_available_versions` so don't hesitate to revert that one. It's weird that there is not function like `min_with` that would take a `F: Fn(&Self::Item) -> impl Ord` or similar.

**Comment by Eh2406** [comment]
Dose `min_by_key` work here?

**Comment by mpizenberg** [comment]
Just after that comment I realized the function exists XD and is called `min_by_key`.

**Review by mpizenberg — APPROVED** [review]

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs]: AFAIK panicking should be done on internal library bugs, which the user can do nothing about. There are a few we define as Failures currently, which I wanted to change to panics in some future PR.

As for user introduced bugs... I don't know actually what the consensus is. I just changed the error to be more specific in the latest commit for now, but I'm open to the discussion on the topic.

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs:257]: @Eh2406 
> Do you want to keep the list_available_versions?

@mpizenberg 
> I don't think we should keep list_available_versions in the interface.

Why not? What if our interface is 3 functions:
1) `make_decision()`
2) `list_available_versions()`
3) `get_dependencies()`

The first one has default implementation in terms on `list_available_versions`, no need to change it right away for the algorithm to work but the ability to customize it is there.

The idea is somewhat similar to what iterators to with the `next()` function. You define it and get some others for free. Here we would implement 1) in terms of 2).

The only not so pretty thing about that is if the user comes up with a way to redefine `make_decision` without using `list_available_versions`, but the API still requires it. In that case they could leave it `unimplemented!`. Is it an API smell? Hmmm.

**Comment by aleksator** [comment]
I've reviewed the PR so far, great work introducing the ability to customize the lib's behavior.

The one thing I wanted to discuss about the API change is [a default implementation for `make_decision`](https://github.com/pubgrub-rs/pubgrub/pull/50#discussion_r523457875). We can leave that discussion for a potential separate PR if you'd like.

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs:230]: I forgot about `should_cancel` but the point stands.

File: tests/proptest.rs
Comment (aleksator) [inline tests/proptest.rs:25]: Is it still true?

```
        Ok(choose_package_with_fewest_versions(
            v
            |p| self.0.versions(p).into_iter().flatten().cloned(),
        })
            potential_packages,
        ))
```
Does not look like we pick "oldest" versions here anymore :thinking:

File: tests/proptest.rs
Comment (aleksator) [inline tests/proptest.rs:33]: As we note in the docs:
>   /// The strategy employed to choose such package and version
    /// cannot change the existence of a solution or not,

So why should it be possible to return an error here?

What we can do is to return `Option<(T, Option<V>)>` here instead of the result.
Then change `choose_package_with_fewest_versions` to return an `Option` as well and instead of
```
        .expect("potential_packages gave us an empty iterator");
```
we can just return `None`.

Alternatively we get don't introduce the `Option` in this type signature (keeping just `(T, Option<V>)`) and we can leave the above mentioned panic in.

Thoughts?

File: tests/proptest.rs
Comment (aleksator) [inline tests/proptest.rs]: What about naming this `[pick/choose]_next_package`?

**Review by aleksator — CHANGES_REQUESTED** [review]

File: tests/proptest.rs
Comment (mpizenberg) [reply tests/proptest.rs:33]: There are lots of things that can go wrong while giving control back to the implementer of `make_decision` like when reading files on disk. Without a result, we force the implementer to panic inside `make_decision`. With a result, they can handle the error that is propagated to `resolve` however they want.

File: tests/proptest.rs
Comment (mpizenberg) [reply tests/proptest.rs]: Could have been possible. `make_decision` fits well with PubGrub terminology since this step is called ["decision making"](https://github.com/dart-lang/pub/blob/master/doc/solver.md#decision-making)

File: tests/proptest.rs
Comment (Eh2406) [reply tests/proptest.rs:25]: `versions` currently returns them in sorted order, So taking the first will be the smallest/oldest.
`versions` does not document this behavior, but this is only used for testing so I don't mind depending on undocumented behavure.

File: src/solver.rs
Comment (Eh2406) [reply src/solver.rs:257]: It is a workable solution, if we had a lot of users of the current api that would be a wonderful way to add an opt in for performance without breaking existing users. The boilerplate required to use the helper function, did not end up being so onerous IMO. I think it is a code smell to require our users to provide something that we will not be using.
I don't think "avoid the breaking change" justifies that smell  given our user base, nor do I think "make it easy to get started" justifies it with the helper function. But it can work if that is what we want to do.

**Comment by Eh2406** [comment]
Using the benchmark of solving all versions in [elm-packages.ron](https://github.com/pubgrub-rs/pubgrub/issues/44#issuecomment-727289232), criterion thinks this PR makes a 19% improvement from 2.2sec -> 1.8sec.
Solving for all versions in the checked in large_case.ron criterion thinks this PR makes a 23% improvement.

File: tests/proptest.rs
Comment (aleksator) [reply tests/proptest.rs:33]: Makes sense.

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs:257]: > I don't think "avoid the breaking change" justifies that smell given our user base

My motivation is not avoiding breaking changes, but rather the easiness of API usage.

Before this PR it was crystal clear what the algorithm needed from the user to function (from the user's perspective):
1) A way to get dependencies for a given package version.
2) A way to get all versions for the package.

All makes sense why it is needed and how to implement it.

Now:
1) Unchanged.
2) Hm, actually you need to implement part of our algorithm to use our library.

How would people do that? What if they don't want to understand how exactly we do the decision making?

The only practical solution is to copy paste the code from `OfflineDependencyProvider`.
I'm afraid that requiring every user to do that is too much of an ergonomic hit.

Does anyone else share my concerns?

File: tests/proptest.rs
Comment (aleksator) [reply tests/proptest.rs]: That's true. The question is who do we optimize the easiness to understand for:
1) library authors
2) users

I think `make_decision` optimizes for us, while naming function based on what it does optimizes for users.

What if we use descriptive function name from the user perspective and put a link to "decision making" in the function docs?

File: src/solver.rs
Comment (mpizenberg) [reply src/solver.rs:257]: I get your point about ease of use. However, I think people not wanting to use `OfflineDependencyProvider` and willing to define their own dependency provider are willing to read the implementation of an 11-lines function doing nothing more than a function call to a helper function. Also there is an even simpler [6-lines example implementation in the guide](https://pubgrub-rs-guide.netlify.app/pubgrub_crate/custom_dep_provider.html) on the section regarding custom implementation of a dependency provider.

If we get more users, and some of them point `make_decision` as being difficult to implement, I think I'd rather have a two different traits with different ease/control compromises (a bit like the [sandbox/element/document/application variants of an elm app](https://package.elm-lang.org/packages/elm/browser/latest/)) than having a single trait with an ambiguous API.

File: tests/proptest.rs
Comment (mpizenberg) [reply tests/proptest.rs]: Yeah I don't know, tough naming call. If you prefer explicitness from the user point of view, maybe `choose_package_version_within` is the most descriptive name? I don't mind if you go for it, but can you also update the guide then please (PR #45).

Comment (aleksator) [reply tests/proptest.rs]: https://github.com/dart-lang/pub/blob/master/lib/src/solver/version_solver.dart#L320 :smile:

Comment (aleksator) [reply tests/proptest.rs]: I went with `choose_package_version`; how does it look now?

I haven't yet read the guide as I'm focusing on PR reviews first. Will update once I get there :slightly_smiling_face:

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs:257]: > If we get more users, and some of them point make_decision as being difficult to implement, I think I'd rather have a two different traits with different ease/control compromises (a bit like the sandbox/element/document/application variants of an elm app) than having a single trait with an ambiguous API.

Actually that's what my girlfriend suggested as well! :laughing: I was pretty impressed because she's sophomore (2nd year in the uni) and is doing that as a way to change her profession, meaning she has very little experience in programming yet.

In any case, doing that is out of scope of this PR and needs a separate discussion; I'm totally okay merging this version!

**Comment by aleksator** [comment]
@Eh2406 That's a lot! Must be usage of iterators instead of allocating new vectors for `list_available_versions` that was done before?

@Eh2406 @mpizenberg 
No blockers from me anymore, let's merge this if you agree with the naming changes in the latest commit discussed in https://github.com/pubgrub-rs/pubgrub/pull/50#discussion_r523755854.
Feel free to resolve and merge.

There is also an interesting discussion on future possible API changes, where @mpizenberg thought of 2 separate traits for customizability/ease of use: https://github.com/pubgrub-rs/pubgrub/pull/50#discussion_r513702296
This is in no way a blocker for this PR, just something we can keep in mind and possibly implement in the future if we desire to do so. I thought it's pretty clever :smile:

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs:257]: I'm glad we had this discussion as this is an interesting design problem I haven't had to think of before, thank you @Eh2406 and @mpizenberg.

**Review by aleksator — APPROVED** [review]

File: src/solver.rs
Comment (mpizenberg) [inline src/solver.rs]: A bit awkward said like that. We've already determined that any existing package version within the constraint is valid so "that will be tested whether ..." is more confusing in my opinion. We should just stick to say that we choose any existing package version within the provided constraints.

**Review by mpizenberg — CHANGES_REQUESTED** [review]

**Comment by Eh2406** [comment]
> Must be usage of iterators instead of allocating new vectors for list_available_versions that was done before?

I think so. Allocating a BTreeSet (in versions) and Vec (in list_available_versions) just to get a count for all available packages for each decision adds up.

@mpizenberg is better at naming than I, so I will leave final approval to him.

File: src/solver.rs
Comment (mpizenberg) [reply src/solver.rs:257]: you have a wise girlfriend @aleksator  ahah, keep her! XD

File: src/solver.rs
Comment (aleksator) [reply src/solver.rs]: I've restored the wording that was there before the change, just added a link to Pubgrub docs regarding the process.

**Review by mpizenberg — APPROVED** [review]
Alright congrats all on finalizing that PR!

**Comment by mpizenberg** [comment]
@Eh2406 There is a conflict and I'm in the middle of some uncommitted things from the old `dev` branch for my analysis of elm packages. Could you do the merge? Don't hesitate to squash everything before rebase since that means only 1 conflictual commit.

**Comment by Eh2406** [comment]
Did a squash and a rebase. I think nothing was lost in that, but just in case I will leave it for one of you to merge.

**Comment by aleksator** [comment]
There was a conflict because of my merged PR; fixed.

**Pull request merged**
